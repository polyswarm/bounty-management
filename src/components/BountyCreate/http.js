class Http {
  constructor(url) {
    this.url = url;
    this.xhr = null;
  }

  cancel() {
    const { xhr } = this;
    if (xhr) {
      xhr.abort();
    }
  }

  uploadFiles(files) {
    return new Promise((resolve, reject) => {
      const url = this.url+'/artifacts';
      if(this.url && files && files.length > 0) {
        // Add files
        const formData = new FormData();
        files.forEach((file) => {
          formData.append(file.name, file, file.name);
        });

        // open connection
        const xhr = new XMLHttpRequest();

        // attach listeners
        xhr.onerror = () => reject(xhr.statusText);
        xhr.onload = () => resolve(xhr.response);

        // send to server
        xhr.open('post', url);
        xhr.send(formData);
        this.xhr = xhr;
      } else {
        reject('No URL set');
      }
    })
      .then(body => body.json)
      .then(json => json.artifacts);
  }

  uploadBounty(amount, artifacts, duration) {
    const url = this.url + '/bounties';
    return new Promise((resolve, reject) => {
      if(this.url && amount && duration && artifacts && artifacts.length > 0) {
        const bounty = {
          amount: amount,
          duration: duration,
          artifacts: artifacts
        };
        resolve(bounty);
      } else {
        reject('Invalid bounty.');
      }
    })
      .then(bounty => fetch(url, {
        method: 'post',
        body: bounty,
      }))
      .then(response => {
        if (response.ok) {
          return response;
        }
        throw new Error('Error response was not ok');
      })
      .then(response => response.json())
      .then(body => body.guid);
  }
}
export default Http;
