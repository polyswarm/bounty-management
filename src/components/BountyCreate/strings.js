const strings = {
  createBounty: 'Create Bounty',
};
export default strings;
