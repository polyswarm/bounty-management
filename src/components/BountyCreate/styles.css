.Bounty-Create {
  background-color: #fafafa;
  border-radius: 20px;
  display: block;
  margin: 0 auto;
  max-width: 768px;
  position: relative;
}

.Container {
  display: grid;
  grid-column-gap: 28px;
  grid-template-columns: repeat(2, 1fr);
  grid-template-rows: 400px 35px 40px;
  padding: 20px;
}

.Bounty-Create-Error {
  color: red;
  grid-column: 1 / 3 ;
  grid-row: 2;
  width: 100%;
}

.Button {
  grid-column: 1/ 3;
  grid-row: 3;
  margin-top: 0;
  width: 100%;
}
