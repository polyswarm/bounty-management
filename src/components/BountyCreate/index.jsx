// Vendor imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';
// Bounty imports
import DropTarget from '../DropTarget';
import FileList from '../FileList';
import Button from '../Button';
import ModalPassword from '../ModalPassword';
// Component imports
import strings from './strings';
import './styles.css';
import Http from './http';

class BountyCreate extends Component {
  constructor(props) {
    super(props);
    this.state = {
      files: [],
      uploading: false,
      error: null,
    };
    this.onMultipleFilesSelected = this.onMultipleFilesSelected.bind(this);
    this.onFileRemoved = this.onFileRemoved.bind(this);
    this.createBounty = this.createBounty.bind(this);
    this.onClickHandler = this.onClickHandler.bind(this);
    this.onClearAll = this.onClearAll.bind(this);
    this.cancel = this.cancel.bind(this);
    this.onWalletChangeHandler = this.onWalletChangeHandler.bind(this);
  }

  componentDidMount() {
    const { props: { url } } = this;
    this.http = new Http(url);
  }

  render() {
    const { state: { files, uploading, error } } = this;
    const { props: { url, walletList } } = this;
    return(
      <div className='Bounty-Create'>
        <ModalPassword ref={(modal) => this.modal = modal}
          url={url}
          walletList={walletList}
          onWalletChange={this.onWalletChangeHandler} />
        <div className='Container'>
          <FileList files={files}
            clear={this.onClearAll}
            removeFile={this.onFileRemoved}/>
          <DropTarget onFilesSelected={this.onMultipleFilesSelected}/>
          {error && (
            <p className='Bounty-Create-Error'>{error}</p>
          )}
          <Button
            className='Bounty-Create-Upload'
            disabled={uploading || !files || files.length === 0 }
            onClick={this.onClickHandler}>
            {uploading && (strings.cancel)}
            {!uploading && (strings.createBounty)}
          </Button>
        </div>
      </div>
    );
  }

  onMultipleFilesSelected(files) {
    const f = this.state.files.slice();
    const combined = f.concat(files);
    this.setState({ files: combined });
  }

  onFileRemoved(index) {
    const files = this.state.files.slice();
    if (index >= 0 && index < files.length) {
      files.splice(index, 1);
      this.setState({ files: files, error: null, });
    }
  }

  onClickHandler() {
    const { state: { uploading } } = this;
    if (uploading) {
      this.cancel();
    } else {
      this.modal.open();
      // if (!isUnlocked) {
      //   this.modal.open();
      // } else {
      //   this.createBounty();
      // }
    }
  }

  onWalletChangeHandler(store) {
    const { props: { onWalletChange } } = this;
    if (onWalletChange) {
      onWalletChange(store);
    }
    this.createBounty();
  }

  onClearAll() {
    this.setState({ files: [], error: null, });
  }

  cancel() {
    const { http } = this;
    http.cancel();
  }

  createBounty() {
    const { props: { addBounty } } = this;
    const { state: { uploading} } = this;
    const files = this.state.files.slice();

    const http = this.http;
    if (!uploading && files && files.length > 0) {
      this.setState({uploading: true, error: null});
      http.uploadFiles(files)
        .then((artifact) => http.uploadBounty('6250000000000000000', artifact, 300))
        .then(result => {
          if (addBounty) {
            addBounty(result);
          }
        })
        .catch((error) => {
          let errorMessage;
          const { props: { onWalletChange } } = this;
          onWalletChange(false);
          if (!error || error.length === 0) {
            errorMessage = strings.error;
          } else {
            errorMessage = error.message;
          }
          this.setState({error: errorMessage});
        })
        .then(() => this.setState({
          uploading: false,
        }));
    }
  }
}

BountyCreate.propTypes = {
  isUnlocked: PropTypes.bool,
  walletList: PropTypes.array,
  onWalletChange: PropTypes.func,
  addBounty: PropTypes.func,
  url: PropTypes.string,
};
export default BountyCreate;
