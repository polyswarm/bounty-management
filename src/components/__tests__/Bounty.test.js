import React from 'react';
import {render} from 'enzyme';
import {renderToJson} from 'enzyme-to-json';
import BountyCreate from '../BountyCreate';

it('renders without crashing', () => {
  const wrapper = render(<BountyCreate />);
  expect(renderToJson(wrapper)).toMatchSnapshot();
});
