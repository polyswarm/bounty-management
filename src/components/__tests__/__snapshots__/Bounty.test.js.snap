// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  class="Upload"
>
  <div
    class="Container"
  >
    <div
      class="Drop-Target"
    >
      <div>
        <div
          class="center"
        >
          <p>
            Drag and drop your files here, or 
          </p>
          <form>
            <input
              class="hidden"
              id="file"
              type="file"
            >
              
            </input>
            <label
              for="file"
            >
              click here.
            </label>
          </form>
        </div>
      </div>
    </div>
    <div
      class="File-List"
    >
      <header>
        Bounty Files 0
      </header>
      <div
        class="List"
      >
        
      </div>
    </div>
  </div>
</div>
`;
